### DISCUSSION TITLE
CAN wiring to Pylontech battery - share your setup?

### DISCUSSION THREAD

**Discussion by martingro** [hoylabs/OpenDTU-OnBattery#245] [General]
Hi all
Thanks for the awesome Project!
I'am almost done. 6x410W on 3xMPPT 100/20 to the Pylontech US5000. Just the CAN is not working.
Build an Wroom ESP32board with the suggested CAN interface now (had a TJA 1040 but wasn't successfull either). Tried with 3 wires (GND included) and also 2 wire (~1m unshielded old Ethernet Cable, pin 4 and 5 used, all 4 DIPs on 0). Power cycled Battery and ESP afterconnecting. But it doesn't communicate anything. 
Can you share your Setup? Do you have shielded wire?

**Comment by sylvio69** [comment]
Same problem here. I solved it for me. Remove the 120 ohm resistor from can module (SN65HV...) and add a 100nf capacitor to +3.3V and Gnd directly on the can module. Works like a charm with 2m ethernet cable (CAN-H, CAN-L, CAN-Gnd) between US5000 and dtu.

**Reply by Manos1966** [reply]
Thank you @sylvio69 (and @martingro )
I am waiting for the Pylontech to arrive and I believe your answer will spare me from a lot of headaches 👍 

If you could share a picture of what and where you soldered, I would appreciate.
I was so happy to have managed to squeeze everything into a box, now I am afraid I have to find a bigger one...

**Reply by Manos1966** [reply]
We are talking about this module, right? (at the lower right of the box).
![WhatsApp Image 2023-06-13 at 16 48 47](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/129660360/eb7073bf-6726-4c2b-8cea-6cac3896fb72)

**Reply by sylvio69** [reply]
Dein Modul scheint einen Jumper für den Abschlusswiderstand zu haben. Also brauchst Du nur den Jumper ziehen und den 100nF nachrüsten falls notwendig. Prinzipiell ist es ja richtig einen CAN Bus an beiden Enden mit 120 Ohm zu terminieren. Den Tip den Widerstand zu entfernen habe ich aus Mikrocontroller.Net. Dort gab es eine Diskussion über Fake Module. Anscheinend gibt es Module wo statt der 3.3V Chips billigere 5V Chips verlötet sind. Und dann kommen die Probleme. Bei mir gehts immer noch wunderbar ohne Abschlusswiderstand. Wir haben ja auch nur kurze Kabel. Bei meiner Hausbus Installation auf CAN Bus Basis ist der Widerstand an beiden Enden dagegen zwingend notwendig, sonst geht gar nix. Das sind aber auch zig Meter Kabel.

Also ich würde Dir empfehlen erstmal mit Widerstand (gesteckter Jumper) zu probieren. Wenn nichts funktioniert dann ohne. Den Kondensator würde ich auf jeden Fall nachrüsten, steht im Datenblatt.

Anbei Bilder meiner DTU. R2 ist der Abschlusswiderstand.

![1686729820019](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/45991702/d4dabf09-d4bb-4578-b2a3-17b4e2b92dd4)
<img width="319" alt="1" src="https://github.com/helgeerbe/OpenDTU-OnBattery/assets/45991702/7ed28fd2-8497-4e72-9483-85e5b7a3bf52">

**Reply by Manos1966** [reply]
Many thanks!

**Comment by Manos1966** [comment]
@sylvio69 So, finally the Pylontech arrived and I got it working! (OK, my box needs improvement... It is just a test, I could not use the OpenDTU I have installed outside as it is raining today so, I quickly put something together just to see if the CANbus is working)  
Many thanks for your invaluable input :-)
Just one question: Do you need to connect a Ground (GND) to the Pylontech cable or is the CAN-H and CAN-L enough?

![20230721_222939](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/129660360/782a8ab8-a8d4-4280-ba30-b40f1bb27cc5)
![Screenshot_20230721-231804_Samsung Internet](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/129660360/dd150cad-dd6f-4664-9999-71bed203da4b)

**Reply by sylvio69** [reply]
CAN can work with just two wires, but doesn't have to. It is always advantageous to use GND for a common reference potential. So I would recommend 3 wires, CAN-H, CAN-L **and** GND.
Last update 175 seconds ago? Pylontech should send an update every second.

**Reply by Manos1966** [reply]
> Last update 175 seconds ago? Pylontech should send an update every second.

It was just the screenshot... I had already switched off the system.

**Comment by Manos1966** [comment]
@sylvio69 I am impressed how you managed to solder everything within such a small space 🥇 
(my poor eyesight + terrible soldering skills do not allow such precision).
The unit on the right side (red arrow), is an "IIC I2C Logic Level Converter", right? Do you use this only for the Victron connection or have you connected the CANbus module as well?
![Question](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/129660360/67abc386-c0bd-4019-bacb-a548c621384c)

**Reply by sylvio69** [reply]
Yes, this is a simple logic level converter (not I2C) only for Victron MPPT. On the left site is the CAN Transceiver.
I've now built a new dtu with other components in the hope that this runs more stable but without success. Still at least one reboot per day. There I use an ADUM1201 instead of the logic converter. There is also a galvanic isolation. I also installed two of these in the hope of connecting two MPPTs in the future (@helgeerbe :)). By the way, I also need reading glasses and a magnifying glass for such things :).

![IMG_1294](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/45991702/5e53c47d-5ef6-43b4-b4ff-2334c0ac492b)

**Reply by Manos1966** [reply]
"galvanic isolation" yeah, I remember the discussion about that...
I am still in "test-mode", extremely happy that I could read the Pylotech Battery and waiting for a cable set to connect the battery to the Hoymiles. After that, I need to see how connect my Huawei charger :-)

So here is another question I have:
Since the CANbus for the Pylontech battery is working and I get reliable battery data from it, **why do I need to receive data from the Victron?**: 
1. For Information (see what the production of the solar panels is)? Yes, I can see the interest in knowing this, especially if I can connect four PV panels per Victron MPTT 100/20 and get two such sets to charge my battery (two Victrons, eight solar panels =  3,2kW peak. der Winter kann kommen 👍 ).
2. For the Solar Pass-through. But I do not need the Solar Pass-through if I have a battery with BMS and a Victron that works with Pylontech, right? 
**If** I understood correctly, Solar Pass-through is used (when the battery is full already) to channel Victron's electricity production to the electricity network in order to avoid overcharging the battery.  
"Solar Pass-through" is a fantastic system for people who have old batteries without BMS, communication with Victron etc, etc. But in my case, having bought expensive hardware (Victron + Pylontech), I have the impression that Solar Pass-through is not necessary.
**Or am I wrong in my thinking?**

**Reply by sylvio69** [reply]
My understanding is as follows, please correct me... First of all, you have to differentiate between the "Normal Solar Passthrough" and the "Full Solar Passthrough". What you wrote under 2) is the "Full Solar Passthrough". Whether or what you need this mode for, everyone has to decide for themselves. In this mode, everything that is possible, i.e. everything that still comes from the roof, is fed into the grid. One reason for this would be you don't want to keep charging your battery, another reason you want to let your counter run backwards if that's possible (illegally I know) or you want to do something good for your energy supplier or you want to avoid the MPPT going into absorption or floating mode. For whatever reason.
The "Normal Solar Passthrough" mode, as far as I understand it, ensures that the electricity from the roof is primarily used to compensate for the purchase from the grid. Only what is left is loaded into the battery. In addition, the battery is only discharged when the start SOC is reached or nothing comes from the roof (night). However, I don't really know how it behaves without "Solar Passthrough" because I've never tried it.
